File: lakeSide-hotel-demo-client-main/src/model/BookedRoom.js
class BookedRoom {
    constructor(
        id = null,
        bookingDateTime = null,
        checkInDate = null,
        checkOutDate = null,
        guestFullName = null,
        guestEmail = null,
        NumOfAdults = null,
        NumOfChildren = null,
        children5_11 = null,
        totalNumOfGuest = null,
        numOfRoom = null,
        bookingConfirmationCode = null,
        status = null,
        room = null, // Đối tượng của class RoomModel
        services = null, // Mảng các đối tượng của class Service
        user = null, // Đối tượng của class User
        bill = null, // Đối tượng của class Bill
        surcharges = null // Mảng các đối tượng của class Surcharge
    ) {
        this.id = id;
        this.bookingDateTime = bookingDateTime;
        this.checkInDate = checkInDate;
        this.checkOutDate = checkOutDate;
        this.guestFullName = guestFullName;
        this.guestEmail = guestEmail;
        this.NumOfAdults = NumOfAdults;
        this.NumOfChildren = NumOfChildren;
        this.children5_11 = children5_11;
        this.totalNumOfGuest = totalNumOfGuest;
        this.numOfRoom = numOfRoom;
        this.bookingConfirmationCode = bookingConfirmationCode;
        this.status = status;
        this.room = room;
        this.services = services;
        this.user = user;
        this.bill = bill;
        this.surcharges = surcharges;
    }
}

export default BookedRoom;

File: lakeSide-hotel-demo-server-master/src/main/java/com/dailycodework/lakesidehotel/model/BookedRoom.java
package com.dailycodework.lakesidehotel.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;

/**
 * @author Simpson Alfred
 */
@Entity
@Getter
@Setter
@AllArgsConstructor
public class BookedRoom {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private  Long id;

    private LocalDateTime bookingDateTime;

    private LocalDate checkInDate;

    private LocalDate checkOutDate;

    private String guestFullName; 

    private String guestEmail;

    @JsonProperty("NumOfAdults")
    private int NumOfAdults;

    @JsonProperty("NumOfChildren")
    private int NumOfChildren;

    private  int children5_11;

    private int totalNumOfGuest;

    private int numOfRoom;

    private String bookingConfirmationCode;

    private String status;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "room_id")
    private Room room;

    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(name = "used_service",
            joinColumns = @JoinColumn(name = "booking_id", referencedColumnName = "id"),
            inverseJoinColumns = @JoinColumn(name = "service_id", referencedColumnName = "id"))
    private List<Service> services;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;

    @OneToMany(mappedBy="bookedRoom", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Bill> bills;

    @OneToMany(mappedBy="bookedRoom", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Surcharge> surcharges;

    public BookedRoom() {
        this.surcharges = new ArrayList<>();
        this.services = new ArrayList<>();
        this.bills = new ArrayList<>();
    }

    public void calculateTotalNumberOfGuest(){
        this.totalNumOfGuest = this.NumOfAdults + NumOfChildren;
    }

    public void setNumOfAdults(int numOfAdults) {
        NumOfAdults = numOfAdults;
        calculateTotalNumberOfGuest();
    }

    public void setNumOfChildren(int numOfChildren) {
        NumOfChildren = numOfChildren;
        calculateTotalNumberOfGuest();
    }

    public void setBookingConfirmationCode(String bookingConfirmationCode) {
        this.bookingConfirmationCode = bookingConfirmationCode;
    }

    public void AddSurcharges(Surcharge surcharge){
        if(this.surcharges == null){
            this.surcharges = new ArrayList<>();
        }
        this.surcharges.add(surcharge);
        surcharge.setBookedRoom(this);
    }

    public void AddBill(Bill bill){
        if(this.bills == null){
            this.bills = new ArrayList<>();
        }
        this.bills.add(bill);
        bill.setBookedRoom(this);
    }
}
